build-tools/packages/build-cli/src/library/git.ts

```diff
@@ -207,4 +210,36 @@ export class Repository {
 
 		return mergeResult.result === "success";
 	}
+
+	/**
+	 * Returns an array containing repo repo-relative paths to all the files in the provided directory.
+	 * A given path will only be included once in the array; that is, there will be no duplicate paths.
+	 * Note that this function excludes files that are deleted locally whether the deletion is staged or not.
+	 */
+	public async getFiles(): Promise<string[]> {
+		const results = await this.gitClient.raw(
+			"ls-files",
+			// Includes cached (staged) files.
+			"--cached",
+			// Includes other (untracked) files that are not ignored.
+			"--others",
+			// Excludes files that are ignored by standard ignore rules.
+			"--exclude-standard",
+			// Removes duplicate entries from the output.
+			"--deduplicate",
+			// Shows the full path of the files relative to the repository root.
+			"--full-name",
+		);
+		const files = results
+			.split("\n")
+			.map((line) => line.trim())
+			// Deleted files that are _unstaged_ will still be in the ls-files results, so filter out missing files.
+			.filter((file) => {
+				// Use absolute path to ensure consistent behavior regardless of working directory.
+				const absPath = path.resolve(this.baseDir, file);
+				return existsSync(absPath);
+			});
```

Comment:
I don't think I can live with a synchronous one-by-one check of all files in repo.
Use `stat` from `fs/promises` and:
```suggestion
		const allFiles = results.split("\n").map((line) => line.trim());
		// Deleted files that are _unstaged_ will still be in the ls-files results, so filter out missing files.
		const existences = await Promise.all(
			allFiles.map(async (file) => {
				// Use absolute path to ensure consistent behavior regardless of working directory.
				const absPath = path.resolve(this.baseDir, file);
				return stat(absPath)
					.then(() => true)
					.catch(() => false);
			}),
		);
		const files = allFiles.filter((_file, index) => existences[index]);
```

Reply:
(I did not test that suggestion.)

Reply:
I originally used stat; I'll switch back.